crypto/evp/digest.c

```diff
@@ -70,6 +70,42 @@ int EVP_MD_CTX_reset(EVP_MD_CTX *ctx)
     return 1;
 }
 
+EVP_MD_CTX *evp_md_ctx_new_ex(EVP_PKEY *pkey, const ASN1_OCTET_STRING *id,
+                              OPENSSL_CTX *libctx, const char *propq)
+{
+    EVP_MD_CTX *ctx;
+    EVP_PKEY_CTX *pctx = NULL;
+
+    if ((ctx = EVP_MD_CTX_new()) == NULL
+        || (pctx = EVP_PKEY_CTX_new_from_pkey(libctx, pkey, propq)) == NULL) {
+        X509err(0, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+#ifndef OPENSSL_NO_SM2
+# ifndef FIPS_MODULE /* TODO(3.0): EVP_PKEY_CTX_ctrl() not yet supported in FIPS module */
+#  ifndef OPENSSL_NO_EC /* TODO why is EVP_PKEY_CTX_set1_id only available then? */
```

Comment:
`EVP_PKEY_CTX_set1_id` is defined in the ec.h header file. It seems misplaced there to me. It's an EVP function and therefore presumably should be in evp.h. I guess the reason was that there are a few other EVP_* macros in that header which are EC specific and the SM2 authors just followed that practice. To my eyes all of those macros are misplaced there.

Currently EVP_PKEY_CTX_set1_id is only useful for SM2 - which is based on EC so I can see why it might have ended up there...but still it seems wrong. Obviously because this is in ec.h it inherits the "no-ec" aspect of that. That would go away if this was instead defined in evp.h.....but not necessarily something we need to sort out as part of this PR.

Reply:
Thanks for your thoughts and explanation.
Also the documentation suggests that  `#define EVP_PKEY_CTX_set1_id()` and its two siblings should be in evp.h - I've simply moved them there.
Looks like all those ugly guards are not needed any more :)

Reply:
There was still a build issue with no-sm2, which seems to imply no-ec.
Upon a closer look I found that the `#define`s for EVP_PKEY_CTRL_SET1_ID and EVP_PKEY_CTRL_GET1_ID{,_LEN} were wrong and misplaced - fixed in commit c0646ecad65f38c1f4e6a086f8ec92dc9a5834e5